msticpy/data/azure_sentinel.py

```diff
@@ -0,0 +1,209 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""Uses the Azure Python SDK to collect and return details related to Azure."""
+from typing import Dict, List
+
+import pandas as pd
+import requests
+from azure.common.exceptions import CloudError
+
+from .azure_data import AzureData
+
+_PATH_MAPPING = {
+    "ops_path": "/providers/Microsoft.SecurityInsights/operations",
+    "alert_rules": "/providers/Microsoft.SecurityInsights/alertRules",
+    "ss_path": "/savedSearches",
+    "bookmarks": "/providers/Microsoft.SecurityInsights/bookmarks",
+}
+
+
+class AzureSentinel(AzureData):
+    """Class for reutrning key Azure Sentinel elements."""
+
+    def __init__(self, connect: bool = False):
+        """Initialize connector for Azure APIs."""
+        super().__init__()
+
+    def connect(
+        self,
+        auth_methods: List = None,
+        silent: bool = False,
+    ):
+        """Authenticate with the SDK & API."""
+        super().connect(auth_methods=auth_methods, silent=silent)
+        self.token = _get_token(self.credentials)
+        self.res_group_url = None
+        self.prov_path = None
+
+    def get_sentinel_workspaces(self, sub_id: str) -> Dict:
+        """
+        Return a list of Azure Sentinel workspaces in a Subscription.
+
+        Parameters
+        ----------
+        sub_id : str
+            The subscription ID to get a list of workspaces from
+
+        Returns
+        -------
+        Dict
+            A dictionary of workspace names and ids
+
+        """
+        print("Finding Azure Sentinel Workspaces...")
+        res = self.get_resources(sub_id=sub_id)
+        # handle no results
+        if isinstance(res, pd.DataFrame) and not res.empty:
+            sentinel = res[
+                (res["resource_type"] == "Microsoft.OperationsManagement/solutions")
+                & (res["name"].str.startswith("SecurityInsights"))
+            ]
+            workspaces = []
+            for wrkspace in sentinel["resource_id"]:
+                res_details = self.get_resource_details(
+                    sub_id=sub_id, resource_id=wrkspace
+                )
+                workspaces.append(res_details["properties"]["workspaceResourceId"])
+
+            workspaces_dict = {}
+            for wrkspace in workspaces:
+                name = wrkspace.split("/")[-1]
+                workspaces_dict.update({name: wrkspace})
+
+            return workspaces_dict
+
+        print(f"No Azure Sentinel workspaces in {sub_id}")
+        return {}
+
+    def get_hunting_queries(self, res_id) -> pd.DataFrame:
```

Comment:
The resource ID is especially klunky and ripe for user errors.
It might be good to have something that sets the current workspace for the class. This would store sub_id, resource_group, workspace name. Then you could just build the resource string without having to supply a parameter.
You could still have the param available as an override for the default.

Perhaps in the connect function we have optional params for subID and workspace name or ID. Then we set those to self. properties. All of the other methods could then reference these values.

Reply:
Not sure I agree with this, users are going to be accessing these elements in either one of two ways - the first is programmatically where passing a resource ID is easy especially as get_sentinel_workspaces returns it for you:
ws = azs.get_sentinel_workspaces(sub_id)
azs.get_hunting_queries(ws['MyWorkspace'])

The other way is people copy and pasting values into the Notebook, probably from the Azure portal in which case it would see one copy and paste of the resource ID is safer and less error prone than 3 separate copy and pastes.

I'll add support for passing the individual elements through if you like, an option and an option to default to try and get them from config.

Ill think on the idea of passing stuff at connect, my gut feeling is that this isn't great as it implies these value are then set for everything in the class which we don't want as most Sentinel customers have multiple workspaces in multiple subscriptions so will likely want to use this to query multiple of them. If I add an option that if you don't provide arguments for the functions it attempts to get them from config files it will effectively achieve the same thing as if you do have a single workspace you can just let it pull from config.